File: TraversalDemo/src/main/scala/Checker.scala
// Pattern-matching typechecker
object Checker {
  import scala.collection.mutable.Map

  type SymTab = Map[String, Type]

  def typecheck(p: Program): Unit = {
    val symtab = Map[String, Type]()

    for (stmt <- p.stmts) {
      typecheck(stmt, symtab)
    }
  }

  def typecheck(stmt: Statement, symtab: SymTab): Unit = stmt match {
    case Decl(id, typ) => {
      if (symtab.isDefinedAt(id.lexeme)) {
        sys.error(s"Variable ${id.lexeme} already declared")
      } else {
        symtab(id.lexeme) = typ
      }
    }

    case Assign(lhs, rhs) => {
      val lexeme = lhs.lexeme

      if (symtab.isDefinedAt(lexeme)) {
        typecheck(rhs, symtab)
        check(symtab(lexeme), rhs.typ)
      } else {
        sys.error(s"Undefined variable $lexeme")
      }
    }

    case PrintInt(expr) => {
      typecheck(expr, symtab)
      check(INT, expr.typ)
    }

    case PrintReal(expr) => {
      typecheck(expr, symtab)
      check(REAL, expr.typ)
    }
  }

  def typecheck(expr: Expression, symtab: SymTab): Unit = expr match {
    case Add(left, right) => {
      typecheck(left, symtab)
      typecheck(right, symtab)

      (left.typ, right.typ) match {
        case (INT, INT) => expr.typ = INT
        case (REAL, REAL) => expr.typ = REAL
        case _ => sys.error("Type mismatch")
      }
    }

    case Num(lexeme) => {
      if (lexeme.contains('.')) {
        expr.typ = REAL
      } else {
        expr.typ = INT
      }
    }

    case Var(lexeme) => {
      if (symtab.isDefinedAt(lexeme)) {
        expr.typ = symtab(lexeme)
      } else {
        sys.error(s"Undefined variable $lexeme")
      }
    }
  }

  def check(expected: Type, actual: Type): Unit = {
    if (expected != actual) {
      sys.error(s"Type mismatch: expected ${expected.toString}, found ${actual.toString}")
    }
  }
}

File: TraversalDemo/src/main/java/edu/depauw/demo/visitor/Checker.java
package edu.depauw.demo.visitor;

import java.util.HashMap;
import java.util.Map;

import edu.depauw.demo.visitor.ast.*;

public class Checker implements StatementVisitor<Void, Void>, ExpressionVisitor<Void, Void> {
	private Map<String, Type> symtab;
	
	public Checker() {
		this.symtab = new HashMap<>();
	}

	public static void typecheck(Program program) {
		Checker checker = new Checker();
		
		for (Statement stmt : program.getStatements()) {
			stmt.accept(checker, null);
		}
	}

	@Override
	public Void visit(Decl decl, Void t) {
		String lexeme = decl.getId().getLexeme();
		Type type = decl.getType();

		if (symtab.containsKey(lexeme)) {
			throw new RuntimeException("Variable " + lexeme + " already declared");
		} else {
			symtab.put(lexeme, type);
		}
		return null;
	}

	@Override
	public Void visit(Assign assign, Void t) {
		String lexeme = assign.getLhs().getLexeme();
		Expression rhs = assign.getRhs();

		if (symtab.containsKey(lexeme)) {
			rhs.accept(this, t);
			check(symtab.get(lexeme), rhs.getType());
		} else {
			throw new RuntimeException("Undefined variable " + lexeme);
		}
		return null;
	}

	@Override
	public Void visit(PrintInt printInt, Void t) {
		Expression expr = printInt.getExpr();

		expr.accept(this, t);
		check(Type.INT, expr.getType());
		return null;
	}

	@Override
	public Void visit(PrintReal printReal, Void t) {
		Expression expr = printReal.getExpr();

		expr.accept(this, t);
		check(Type.REAL, expr.getType());
		return null;
	}

	@Override
	public Void visit(Add add, Void t) {
		Expression left = add.getLeft();
		Expression right = add.getRight();
		
		left.accept(this, t);
		right.accept(this, t);

		if (left.getType() == Type.INT && right.getType() == Type.INT) {
			add.setType(Type.INT);
		} else if (left.getType() == Type.REAL && right.getType() == Type.REAL) {
			add.setType(Type.REAL);
		} else {
			throw new RuntimeException("Type mismatch");
		}
		return null;
	}

	@Override
	public Void visit(Num num, Void t) {
		String lexeme = num.getLexeme();

		if (lexeme.contains(".")) {
			num.setType(Type.REAL);
		} else {
			num.setType(Type.INT);
		}
		return null;
	}

	@Override
	public Void visit(Var var, Void t) {
		String lexeme = var.getLexeme();

		if (symtab.containsKey(lexeme)) {
			var.setType(symtab.get(lexeme));
		} else {
			throw new RuntimeException("Unknown variable " + lexeme);
		}
		return null;
	}
	
	private static void check(Type expected, Type actual) {
		if (expected != actual) {
			throw new RuntimeException("Type mismatch: expected " + expected + ", found " + actual);
		}
	}
}
